File: easy/1022_sum_of_root_to_leaf_binary_numbers/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <numeric>
#include <vector>

class Solution {
public:
  int sumRootToLeaf(utils::TreeNode *root) {
    std::vector<std::pair<utils::TreeNode *, int>> stack = {{root, root->val}};
    std::vector<int> nums;

    while (!stack.empty()) {
      utils::TreeNode *node = stack.back().first;
      int num = stack.back().second;
      stack.pop_back();

      if (!node->left && !node->right) {
        nums.push_back(num);
        continue;
      }

      if (node->left)
        stack.push_back({node->left, (num << 1) | node->left->val});
      if (node->right)
        stack.push_back({node->right, (num << 1) | node->right->val});
    }

    return reduce(nums.begin(), nums.end(), 0);
  }
};

File: easy/1022_sum_of_root_to_leaf_binary_numbers/solution.py
from typing import Optional

from utils.python.nodes import TreeNode


class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        stack = [(root, [root.val])]
        bin_nums = []

        while stack:
            node, bin_num = stack.pop()

            if not node.left and not node.right:
                bin_nums.append(bin_num)
                continue

            if node.left:
                stack.append((node.left, bin_num + [node.left.val]))
            if node.right:
                stack.append((node.right, bin_num + [node.right.val]))

        nums_sum = 0

        for bin_num in bin_nums:
            num = 0

            for bit in bin_num:
                num = (num << 1) | bit

            nums_sum += num

        return nums_sum
